auth/postgres/repo.go

```diff
@@ -0,0 +1,465 @@
+		SET revoked = true, revoked_at = :revoked_at
+		WHERE user_id = :user_id AND id = :id`
+
+	upm := dbPatPagemeta{
+		User:      userID,
+		ID:        patID,
+		ExpiresAt: time.Now(),
+	}
+
+	_, err := pr.db.NamedQueryContext(ctx, q, upm)
+	if err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) Reactivate(ctx context.Context, userID, patID string) error {
+	q := `
+		UPDATE pats 
+		SET revoked = false, revoked_at = NULL
+		WHERE user_id = :user_id AND id = :id`
+
+	upm := dbPatPagemeta{
+		User: userID,
+		ID:   patID,
+	}
+
+	_, err := pr.db.NamedQueryContext(ctx, q, upm)
+	if err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) Remove(ctx context.Context, userID, patID string) error {
+	q := `DELETE FROM pats WHERE user_id = :user_id AND id = :id`
+	upm := dbPatPagemeta{
+		User: userID,
+		ID:   patID,
+	}
+
+	_, err := pr.db.NamedQueryContext(ctx, q, upm)
+	if err != nil {
+		return postgres.HandleError(repoerr.ErrRemoveEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) AddScopeEntry(ctx context.Context, userID string, scopes []auth.Scope) error {
+	scs := toDBScope(scopes)
+	_, err := pr.db.NamedQueryContext(ctx, saveScopeQuery, scs)
+	if err != nil {
+		return postgres.HandleError(repoerr.ErrUpdateEntity, err)
+	}
+
+	if err := pr.cache.Save(ctx, scopes); err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) RemoveScopeEntry(ctx context.Context, userID string, scopes []auth.Scope) error {
+	dbscopes := toDBScope(scopes)
+
+	_, err := pr.db.NamedQueryContext(ctx, deleteScopesQuery, dbscopes)
+	if err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	if err := pr.cache.Remove(ctx, scopes); err != nil {
+		return errors.Wrap(repoerr.ErrRemoveEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) CheckScopeEntry(ctx context.Context, userID, patID string, entityType auth.EntityType, optionalDomainID string, operation auth.Operation, entityID string) error {
+	authorized, err := pr.cache.CheckScope(ctx, patID, optionalDomainID, entityType, operation, entityID)
+	if err == nil && authorized {
+		return nil
+	}
+
+	scopes, err := pr.retrieveScopeFromDB(ctx, patID)
+	if err != nil {
+		return err
+	}
+
+	for _, sc := range scopes {
+		if sc.Authorized(entityType, optionalDomainID, operation, entityID) {
+			return nil
+		}
+	}
+	return repoerr.ErrNotFound
+}
+
+func (pr *patRepo) RemoveAllScopeEntry(ctx context.Context, userID, patID string) error {
+	scopes, err := pr.retrieveScopeFromDB(ctx, patID)
+	if err != nil {
+		return err
+	}
+
+	scs := toDBScope(scopes)
+
+	_, err = pr.db.NamedQueryContext(ctx, deleteAllScopesQuery, scs)
+	if err != nil {
+		return postgres.HandleError(repoerr.ErrUpdateEntity, err)
+	}
+
+	scopes, err = pr.retrieveScopeFromDB(ctx, patID)
+	if err != nil {
+		return err
+	}
+
+	if err := pr.cache.Save(ctx, scopes); err != nil {
```

Comment:
Please create function `RemoveAllPATScope` in cache, this function will remove all the scope in cache. 
In Redis delete all keys with prefix of PAT ID.